File: Online-judge/leetcode/number-of-operations-to-make-network-connected.py
from collections import defaultdict, deque
from typing import List


# DFS
class Solution:
    def makeConnected(self, n: int, connections: List[List[int]]) -> int:
        if len(connections) < n-1:
            return -1
        networks = defaultdict(list)

        for connection in connections:
            networks[connection[0]].append(connection[1])
            networks[connection[1]].append(connection[0])
        
        def dfs(node, networks, visited):
            visited[node] = True
            for connectedNode in networks[node]:
                if not visited[connectedNode]:
                    dfs(connectedNode, networks, visited)

        disconnectedNetwork = 0
        visited = [False] * n
        for node in range(n):
            if not visited[node]:
                dfs(node, networks, visited)
                disconnectedNetwork += 1
        return disconnectedNetwork - 1
    

# BFS    
class Solution:
    def makeConnected(self, n: int, connections: List[List[int]]) -> int:
        if len(connections) < n-1:
            return -1
        networks = defaultdict(list)

        for connection in connections:
            networks[connection[0]].append(connection[1])
            networks[connection[1]].append(connection[0])
        
        def bfs(node, networks, visited):
            visited[node] = True
            q = deque([node])

            while q:
                top = q.popleft()
                for connectedNode in networks[top]:
                    if not visited[connectedNode]:
                        visited[connectedNode] = True
                        q.append(connectedNode)


        disconnectedNetwork = 0
        visited = [False] * n
        for node in range(n):
            if not visited[node]:
                bfs(node, networks, visited)
                disconnectedNetwork += 1
        return disconnectedNetwork - 1


        
        

File: Online-judge/leetcode/number-of-operations-to-make-network-connected.cpp
# include <bits/stdc++.h>
using namespace std;

/**
 * @author Md.Shohag <shohag.fks@gmail.com>
 * Problem : https://leetcode.com/problems/number-of-operations-to-make-network-connected/description/
 * Time complexity: O(n+e)
 * Space complexity: O(n+e)
*/


class Solution {
private:
    void dfs(int node, vector<vector<int>>& networks, vector<bool>& visited) {
        visited[node] = true;
        for (int connectedNode : networks[node]) {
            if (!visited[connectedNode]) {
                dfs(connectedNode, networks, visited);
            }
        }
    }
    
public:
    int makeConnected(int n, vector<vector<int>>& connections) {
        if (connections.size() < n-1) return -1;
        vector<vector<int>> networks(n);

        for (auto& connection : connections) {
            networks[connection[0]].push_back(connection[1]);
            networks[connection[1]].push_back(connection[0]);
        }
        
        vector<bool> visited(n + 1, false);
        int disconnectedNode = 0;

        for (int node = 0; node < n; node++) {
            if (!visited[node]) {
                dfs(node, networks, visited);
                disconnectedNode++;
            }
        }
        return disconnectedNode - 1;
    }
};